numpyro/infer/hmc_gibbs.py

```diff
@@ -787,7 +789,7 @@ def log_likelihood(params_flat, subsample_indices=None):
                             log_lik[frame.name] += _sum_all_except_at_dim(
                                 site["fn"].log_prob(site["value"]), frame.dim
                             )
-                        else:
+                        elif frame.name in subsample_indices:
```

Comment:
That makes sense. 
I think you could merge the conditionals with a `defaultdict`. Something like
``` python
log_lik = defaultdict(float)
for site in tr.values():
  if site["type"] == "sample" and site["is_observed"]:
    for frame in site["cond_indep_stack"]:
      if frame.name in log_lik or frame.name in subsample_indices:
        log_lik[frame.name] += _sum_all_except_at_dim(
          site["fn"].log_prob(site["value"]), frame.dim)
```

Reply:
I think this works but we need to maintain the type of the returned values. If we use defaultdict here, we will need to use defaultdict in other places like in gibbs_update function below.